lib/coingecko/coingecko_repository.dart

```diff
@@ -0,0 +1,82 @@
+import 'dart:convert';
+
+import 'dart:developer';
+import 'package:decimal/decimal.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:http/http.dart' as http;
+
+import '../../../settings/available_currency.dart';
+import 'coingecko_types.dart';
+
+abstract class CoinGeckoRepository {
+  Future<CoinGeckoPrice> fetchPrice(AvailableCurrency fiat);
+  Future<CoinGeckoRates> fetchRates();
+}
+
+final coinGeckoRepositoryProvider = Provider<CoinGeckoRepository>((_) => CoinGeckoRepositoryImpl(http.Client()));
```

Comment:
Agreed it is a workaround to provide it as an Singleton to the app context. I will change it.

Reply:
I checked multiple ways. It seems that there is no better way to do this without integrating a new library. As far as I can see it this code is the best way to attach the instance to the flutter app lifecycle and avoids multiple initializations of the same object.  

I am just worrying about the http client which usually holds some state regarding the previous connections, like dns entries and keeps connections open to the host. Creating multiple clients could lead to a resource leak. 

In my opinion, despite it is not perfect, it seems to me as the cleanest and safest solution in the current context. I am open for better solutions could not find any without adding an additional framework/library though.

Reply:
ok checked a little bit more and on the the riverpod doc they even mention that it is the recommended way to expose repositories and http clients.

```
Provider
Provider is the most basic of all providers. It creates a value... And that's about it.

Provider is typically used for:

caching computations
exposing a value to other providers (such as a Repository/HttpClient).
offering a way for tests or widgets to override a value.
reducing rebuilds of providers/widgets without having to use select.

```


https://riverpod.dev/docs/providers/provider

Reply:
Yeah, sorry, what I meant is that you should move the `coinGeckoRepositoryProvider` to coingecko_providers.dart. I prefer to have providers in a *_providers.dart file and the repository files in general should not import state management dependency like riverpod.

Reply:
I created an fake "factory method" to avoid the http client import in the providers file. It is more or less an implementation detail related to the repository. Any way I do not want to overthink this. I am fine with initializing it in the providers as well.